rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,295 @@
+Glides will provide high-level trainsheet information as CloudEvents in a Kinesis stream
+
+# Motivation
+As part of digitizing the currently paper-based trainsheets that define operational conditions on the Green Line, Glides will be gathering information about consist assignments, dropped/added trips, and departure time adjustments from its users: pull-out inspectors at selected terminal stations. This data can serve to improve the rider-facing predictions produced by RTR, especially for departure times at a terminals, which currently do not have departure predictions at all.
+
+# Guide-level explanation
+The Glides trainsheet functionality hews closely to what the paper-based trainsheets provide.
+
+- sign an operator in, marking them fit-for-duty
+- add a trip (1-way from the current terminal, 1-way arriving at the current terminal, or a round-trip from the current terminal)
+- drop a trip
+- add/change/remove train cars to a trip
+- add/change/remove the operator assigned to a trip
+- set a new departure time for a trip
+
+These actions are always taken by a logged-in user (generally an inspector).
+
+## CloudEvents
+All events will be in the [CloudEvents v1.0.2](https://github.com/cloudevents/spec/blob/v1.0.2/cloudevents/spec.md) format (or later), using the JSON encoding. The event types will be under the `com.mbta.ctd.glides` event namespace, and will be in the past-tense (i.e. `trip-added` rather than `add-trip`).
+
+If multiple events happen as a part of the same user action, they can be written as a Batch event, wrapping all of the events in a JSON array.
+
+# Kinesis
+Events will be written to a Kinesis stream. Each Glides environment (`dev`, `dev-green`, and `prod`) will have a separate stream for their events. The partition key will be a hash of the station at which the inspector is working and the inspector's identity, which ensures that multiple events from a single inspector are ordered correctly if the messages are distributed across multiple shards.
+
+## Value types
+### Trip Key
+For scheduled trips, Glides does not have the same trip ID used by RTR. For added trips, while an ID is provided, it can be the case that an added trip from the Glides perspective is matches a GTFS trip, in which case RTR will us the GTFS trip ID in publishing the information.
+
+In order to reduce message duplication, a Trip Key is used to identify both added and scheduled trips.
+
+| Added Trip | Scheduled Trip |
+| ---------- | -------------- |
+| `{"serviceDate": "2023-01-20", "id": "ADDED-123"}`|`{"serviceDate": "2023-01-20", "scheduledTime": "01:02:30"}`|
+
+### Trip Type
+When adding a trip, there are three kinds of trips which can be added:
+- `"arrival"`: a trip arriving at `stop` at `time`
+- `"departure"`: a trip departing from `stop` at `time`
+- `"roundTrip"`: a trip departing from `stop` at `time` and expected to return to `stop`
+
+### Actor
+Each message will include an `actor` key, indicating who performed the given intervention.
```

Comment:
I'm not sure I would understand what "actor" meant without reading this document. Would a different term like "supervisor" or "inspector" be easier to figure out?

Reply:
I'm not sure that it's always a supervisor or inspector doing these things, hence the more generic name. I'm open to other ideas, though.

Reply:
I like `author`, so I've made the update to use that language.